Kokkos package failing at install step

## Summary

When trying to build and install the latest stable and patch releases using the CMake procedure, I'm getting failures on the install step related to kokkos.

## Type of Issue

Bug report

## LAMMPS Version (Bug Report)

Tested with latest stable and patch releases

## Expected Behavior (Bug Report)

Successful installaton

## Actual Behavior (Bug Report)

```
-- Final kokkos settings variable:
--   env;KOKKOS_SRC_PATH=/dev/shm/ocaisa/LAMMPS/18Sep2018/intel-para-2018a-kokkos/lammps-patch_18Sep2018/lib/kokkos;KOKKOS_PATH=/dev/shm/ocaisa/LAMMPS/18Sep2018/intel-para-2018a-kokkos/lammps-patch_18Sep2018/lib/kokkos;KOKKOS_INSTALL_PATH=/usr/local/software/jureca/Stages/Devel-2018a/software/LAMMPS/18Sep2018-intel-para-2018a-kokkos;KOKKOS_ARCH=HSW;KOKKOS_DEVICES=OpenMP,Serial;KOKKOS_DEBUG=no;KOKKOS_OPTIONS=disable_dualview_modify_check,enable_eti;KOKKOS_USE_TPLS=librt
-- *****************************************************
-- 
/dev/shm/ocaisa/LAMMPS/18Sep2018/intel-para-2018a-kokkos/lammps-patch_18Sep2018/lib/kokkos/cmake/Makefile.generate_cmake_settings:7: /usr/local/software/jureca/Stages/Devel-2018a/software/LAMMPS/18Sep2018-intel-para-2018a-kokkos/core/src/Makefile.generate_header_lists: No such file or directory
/dev/shm/ocaisa/LAMMPS/18Sep2018/intel-para-2018a-kokkos/lammps-patch_18Sep2018/lib/kokkos/cmake/Makefile.generate_cmake_settings:8: /usr/local/software/jureca/Stages/Devel-2018a/software/LAMMPS/18Sep2018-intel-para-2018a-kokkos/core/src/Makefile.generate_build_files: No such file or directory
make[1]: *** No rule to make target `/usr/local/software/jureca/Stages/Devel-2018a/software/LAMMPS/18Sep2018-intel-para-2018a-kokkos/core/src/Makefile.generate_build_files'.  Stop.
CMake Error at ../lib/kokkos/CMakeLists.txt:43 (message):
  Kokkos settings generation failed:
```
It seems to be looking for the files `Makefile.generate_header_lists` and `Makefile.generate_build_files` in `KOKKOS_INSTALL_PATH` rather than `KOKKOS_SRC_PATH`.

## Steps to Reproduce (Bug Report)

My setup is with the Intel compiler, an MPICH-based MPI, MKL and the required deps for the following CMake command:
```
cmake ./cmake -DCMAKE_INSTALL_PREFIX=/usr/local/software/jureca/Stages/Devel-2018a/software/LAMMPS/18Sep2018-intel-para-2018a-kokkos -DCMAKE_C_COMPILER='mpicc' -DCMAKE_Fortran_FLAGS='-O2 -xHost -ftz -fp-speculation=safe -fp-model source -fopenmp' -DCMAKE_CXX_FLAGS='-O2 -xHost -ftz -fp-speculation=safe -fp-model source -fopenmp' -DCMAKE_CXX_COMPILER='mpicxx' -DCMAKE_Fortran_COMPILER='mpif90' -DCMAKE_C_FLAGS='-O2 -xHost -ftz -fp-speculation=safe -fp-model source -fopenmp' -DCMAKE_VERBOSE_MAKEFILE=ON -DBUILD_MPI=on -DBUILD_OMP=on -DPKG_KOKKOS=on -DPKG_OPT=on -DPKG_USER-OMP=on -DPKG_USER-INTEL=on -DPKG_GPU=on -DPKG_ASPHERE=on -DPKG_BODY=on -DPKG_CLASS2=on -DPKG_COLLOID=on -DPKG_COMPRESS=on -DPKG_CORESHELL=on -DPKG_DIPOLE=on -DPKG_GRANULAR=on -DPKG_KSPACE=on -DPKG_MANYBODY=on -DPKG_MC=on -DPKG_MEAM=on -DPKG_MISC=on -DPKG_MOLECULE=on -DPKG_PERI=on -DPKG_QEQ=on -DPKG_REAX=on -DPKG_REPLICA=on -DPKG_RIGID=on -DPKG_SHOCK=on -DPKG_SNAP=on -DPKG_SRD=on -DPKG_PYTHON=on -DPKG_MPIIO=on -DPKG_POEMS=on -DPKG_USER-ATC=on -DPKG_USER-AWPMD=on -DPKG_USER-CGDNA=on -DPKG_USER-CGSDK=on -DPKG_USER-COLVARS=on -DPKG_USER-DIFFRACTION=on -DPKG_USER-DPD=on -DPKG_USER-DRUDE=on -DPKG_USER-EFF=on -DPKG_USER-FEP=on -DPKG_USER-H5MD=on -DPKG_USER-LB=on -DPKG_USER-MANIFOLD=on -DPKG_USER-MEAMC=on -DPKG_USER-MESO=on -DPKG_USER-MGPT=on -DPKG_USER-MISC=on -DPKG_USER-MOFFF=on -DPKG_USER-MOLFILE=on -DPKG_USER-NETCDF=on -DPKG_USER-PHONON=on -DPKG_USER-QTB=on -DPKG_USER-REAXC=on -DPKG_USER-SMTBQ=on -DPKG_USER-SPH=on -DPKG_USER-TALLY=on -DPKG_USER-UEF=on -DBUILD_LIB=on -DBUILD_EXE=on -DBUILD_SHARED_LIBS=on -DWITH_JPEG=yes -DWITH_PNG=yes -DWITH_GZIP=yes -DFFT=MKL -DFFT_PACK=array -DGPU_API=cuda -DGPU_ARCH=sm_30 -DCUDPP_OPT=on -DKOKKOS_ENABLE_OPENMP=yes -DKOKKOS_ARCH="HSW"
```
To reproduce just `make; make install`

## Further Information, Files, and Links

I'm using [EasyBuild](https://easybuild.readthedocs.io/en/latest/) to do the installation, a working easyconfig without Kokkos for this build can be seen [here](https://gist.github.com/ocaisa/965ee383b4ec156b7e9fc0251411d8ec) (but is specific to the setup at Juelich Supercomputing Centre)

--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -69,6 +69,8 @@ get_lammps_version(${LAMMPS_SOURCE_DIR}/version.h LAMMPS_VERSION)
 # Cmake modules/macros are in a subdirectory to keep this file cleaner
 set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/Modules)
 
+include(PreventInSourceBuilds)
+
 if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CXX_FLAGS)
   #release comes with -O3 by default
   set(CMAKE_BUILD_TYPE Release CACHE STRING "Choose the type of build, options are: None Debug Release RelWithDebInfo MinSizeRel." FORCE)

--- a/cmake/Modules/PreventInSourceBuilds.cmake
+++ b/cmake/Modules/PreventInSourceBuilds.cmake
@@ -1,0 +1,22 @@
+# - Prevent in-source builds.
+# https://stackoverflow.com/questions/1208681/with-cmake-how-would-you-disable-in-source-builds/
+
+function(prevent_in_source_builds)
+  # make sure the user doesn't play dirty with symlinks
+  get_filename_component(srcdir "${CMAKE_SOURCE_DIR}" REALPATH)
+  get_filename_component(srcdir2 "${CMAKE_SOURCE_DIR}/.." REALPATH)
+  get_filename_component(bindir "${CMAKE_BINARY_DIR}" REALPATH)
+
+  # disallow in-source builds
+  if("${srcdir}" STREQUAL "${bindir}" OR "${srcdir2}" STREQUAL "${bindir}")
+    message(FATAL_ERROR "\
+
+CMake must not to be run in the source directory. \
+Rather create a dedicated build directory and run CMake there. \
+To clean up after this aborted in-place compilation:
+  rm -r CMakeCache.txt CMakeFiles
+")
+  endif()
+endfunction()
+
+prevent_in_source_builds()

